### ISSUE TITLE
HttpClient UseDefaultCredentials Get with kerberos auth server sends NTLM

### ISSUE THREAD

**Issue by NathanielACollier** [dotnet/runtime#51701]
### Description

With `netcoreapp3.1` the following code will send `Kerberos` to the web server.  Switching the application to `.net 5.0` causes `NTLM` to be sent.  The web server is a IBM i Apache, with kerberos configured.  The server does not have NTLM, so the code fails.

### Configuration

Running .net 5.0 on a Windows 10 machine
Connecting to an IBM i Apache Web server with kerberos enabled

### Regression?

This code works if you switch the project to build `netcoreapp3.1` and it fails if you switch it to `.net 5.0`

### Other information

Here is the code that fails

```c#
        static async Task Main(string[] args)
        {
            var url = "https://myIBMi.myDomain.com/api.php?task=getUser";

            var http = new System.Net.Http.HttpClient(new HttpClientHandler()
            {
                UseDefaultCredentials = true
            });

            var result = await http.GetAsync(url);
            string responseText = await result.Content.ReadAsStringAsync();
            
            Console.WriteLine(responseText);
        }
    }
```

## `netcoreapp3.1` calling the code
+ Using fiddler I've captured the raw request and response
+ Initial Request
```
GET https://myIBMi.myDomain.com/api.php?task=getUser
HTTP/1.1
Host: myIBMi.myDomain.com:451

```
+ Initial Response
```
HTTP/1.1 401 Unauthorized
Date: Thu, 22 Apr 2021 17:58:09 GMT
Server: Apache
WWW-Authenticate: Negotiate
Content-Length: 205
Connection: close
Content-Type: text/html; charset=UTF-8
Proxy-Support: Session-Based-Authentication

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<HTML><HEAD>
<TITLE>401 Authorization Required</TITLE>
</HEAD><BODY>
<H1>Authorization Required</H1>
Unauthorized - authentication failed.
</body></html>

```
+ Request to the unauthorized response
     + I abreviated the kerberos ticket information below
```
GET https://myIBMi.myDomain.com/api.php?task=getUser
HTTP/1.1
Host: myIBMi.myDomain.com:451
Authorization: Negotiate YIII8gYGKwYBB....

```
+ Then the response is a success.  Not sure it's important to show that part.

## change the project to `.net 5.0`
+ Initial Request
```
GET https://myIBMi.myDomain.com/api.php?task=getUser
HTTP/1.1
Host: myIBMi.myDomain.com:451

```
+ Initial Response
```
HTTP/1.1 401 Unauthorized
Date: Thu, 22 Apr 2021 18:10:57 GMT
Server: Apache
WWW-Authenticate: Negotiate
Content-Length: 205
Connection: close
Content-Type: text/html; charset=UTF-8
Proxy-Support: Session-Based-Authentication

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<HTML><HEAD>
<TITLE>401 Authorization Required</TITLE>
</HEAD><BODY>
<H1>Authorization Required</H1>
Unauthorized - authentication failed.
</body></html>

```
+ Request to respond to initial response
     + I have edited it to not show the full ntlm value
```
GET https://myIBMi.myDomain.com/api.php?task=getUser 
HTTP/1.1
Host: myIBMi.myDomain.com:451
Authorization: Negotiate TlRMTVNTUA....

```

**Comment by xqrzd** [comment]
I've also had this same problem with .NET 5. HttpClient Kerberos authentication works for IIS sites, but fails for PrestoDB and WebHDFS. Both work in .NET Core 3.1.

**Comment by mrducnguyen** [comment]
I'm having the same problem I think

The Project I'm working on is consuming REST API from Objective ECM system. It's been working using default credential authentication on **.Net Core 3.1**

When I changed the project to target **.Net 5.0**, the authentication failed.

Below is the details request/response headers (using Fiddler to capture them)

### Targeting `.netcoreapp3.1` 
Initial request
```
GET https://aaaa/api/resources/folders/fA702245 HTTP/1.1
Host: aaaa
Accept: application/json, */*
```
Initial response
```
HTTP/1.1 401 Unauthorized
Expires: 0
Cache-Control: no-cache, no-store, must-revalidate
Set-Cookie: JSESSIONID=aaaa; path=/api; secure; HttpOnly
Server: Objective Application Server 10.5.4.3(84468)
Pragma: no-cache
Date: Mon, 26 Apr 2021 01:46:33 GMT
Connection: keep-alive
WWW-Authenticate: Negotiate
Strict-Transport-Security: max-age=300; includeSubDomains
Content-Type: text/html;charset=ISO-8859-1
Content-Length: 71
Proxy-Support: Session-Based-Authentication
```

Subsequence request
```
GET https://aaaa/api/resources/folders/fA702245 HTTP/1.1
Host: aaaa
Accept: application/json, */*
Authorization: Negotiate YIIHPAYGKw...
Cookie: JSESSIONID=aaaa
```

Subsequence response
```
HTTP/1.1 200 OK
Expires: 0
Cache-Control: no-cache, no-store, must-revalidate
Server: Objective Application Server 10.5.4.3(84468)
Date: Mon, 26 Apr 2021 01:46:33 GMT
Connection: keep-alive
```

### Targeting `.net5.0`
Initial request
```
GET https://aaaa/api/resources/folders/fA702245 HTTP/1.1
Host: aaaa
Accept: application/json, */*
```
Initial response
```
HTTP/1.1 401 Unauthorized
Expires: 0
Cache-Control: no-cache, no-store, must-revalidate
Set-Cookie: JSESSIONID=aaaa; path=/api; secure; HttpOnly
Server: Objective Application Server 10.5.4.3(84468)
Pragma: no-cache
Date: Mon, 26 Apr 2021 01:46:33 GMT
Connection: keep-alive
WWW-Authenticate: Negotiate
Strict-Transport-Security: max-age=300; includeSubDomains
Content-Type: text/html;charset=ISO-8859-1
Content-Length: 71
Proxy-Support: Session-Based-Authentication
```

Subsequence request
```
GET https://aaaa/api/resources/objects/fA702245 HTTP/1.1
Host: aaaa
Accept: application/json, */*
Authorization: Negotiate TlRMTVNTUAABAAAAl4II4gAAAAAAAAAAAAAAAAAAAAAKADk4AAAADw==
Cookie: JSESSIONID=aaaa
```

Subsequence response
```
HTTP/1.1 401 Unauthorized
Expires: 0
Cache-Control: no-cache, no-store, must-revalidate
Server: Objective Application Server 10.5.4.3(84468)
Pragma: no-cache
Date: Mon, 26 Apr 2021 01:47:11 GMT
Connection: keep-alive
WWW-Authenticate: Negotiate
Strict-Transport-Security: max-age=300; includeSubDomains
Content-Type: text/html;charset=ISO-8859-1
Content-Length: 71
Proxy-Support: Session-Based-Authentication
```

I did a lot of searching and it seems like `HttpClient` was trying to do the handshaking with the server, using NTLMSSP and looked like the server didn't support that.

Is there any workaround in this situation? As we have absolutely no control on the server side to change anything...

**Comment by danmoseley** [comment]
@wfurt

**Comment by wfurt** [comment]
can you check what SPN is used? This would happen when there is SPN mismatch. The decision really comes from GSSAPI. Some of the example above suggest you have non standard port? `Host: myIBMi.myDomain.com:451` ? Make sure you have proper SPN for that. 

The other part to check would be DNS name. Is your `myIBMi.myDomain.com` `A` record or is it CNAME pointing to some other name?

**Comment by mrducnguyen** [comment]
> can you check what SPN is used? This would happen when there is SPN mismatch. The decision really comes from GSSAPI. Some of the example above suggest you have non standard port? `Host: myIBMi.myDomain.com:451` ? Make sure you have proper SPN for that.
> 
> The other part to check would be DNS name. Is your `myIBMi.myDomain.com` `A` record or is it CNAME pointing to some other name?

The host and port of the server I'm accessing is definitely not standard. Full URL looks like this: `https://edrms-pre:8443/api/resources`

Dummy question, how to check for what SPN is used?  Is there any where I can read more on how Authentication is decided by `HttpClient` when `UseDefaultCredentials` is set to `true`? Thanks in advance!

Also, I'm curious why is there a completely different (breaking) behaviour moving to .Net Core 5.0? Is that because .Net Core 5.0 enforcing better standards?

**Comment by wfurt** [comment]
I think you may start with checking the configured SPNs: https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2012-r2-and-2012/cc731241(v=ws.11)
`setspn -l edrms-pre`

To answer the second part: I think this is related to #36644. 3.1 was always wrong not using port and that was fixed in 5.0 to match .NET Framework. That may show up as breaking change in 5.0 if the SPN was not setup properly. 

I don't know if this is also answer for original issue posted bu @NathanielACollier but I think it will impact your case @mrducnguyen. 

If this is the root cause it should be easy to fix by adding port specific SPN to your domain.

**Comment by NathanielACollier** [comment]
> I think you may start with checking the configured SPNs: https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2012-r2-and-2012/cc731241(v=ws.11)
> `setspn -l edrms-pre`
> 
> To answer the second part: I think this is related to #36644. 3.1 was always wrong not using port and that was fixed in 5.0 to match .NET Framework. That may show up as breaking change in 5.0 if the SPN was not setup properly.
> 
> I don't know if this is also answer for original issue posted bu @NathanielACollier but I think it will impact your case @mrducnguyen.
> 
> If this is the root cause it should be easy to fix by adding port specific SPN to your domain.

It is the case that we also have a non standard port 8452 and 8251 that we use for https.

We don't have experience with SPN, and have had trouble with that, and can't justify setting them.  We are going to figure out a work around with curl.  Curl is still able to make the kerberos calls, and there is a CurlThin library that works really well so we are going to use it to make http calls when the SPNs aren't correct.

**Comment by NathanielACollier** [comment]
We got a curl httpclient thrown together, and it works, but people did decide we should engage a consultant to figure out the SPN.  I'm going to write back once we've got a consultant to setup the SPNs with those ports.

**Comment by mrducnguyen** [comment]
Thanks @wfurt for your explanation :)

In our case, the IT infrastructure is completely outsourced to a 3rd party. We'll need to submit a request to add the SPN with port number. It might take days before it is executed. It might need to go through a couple rounds of meetings and discussions. I'll probably need to answer a couple questions before it could be done:  Why do I need to upgrade to .Net 5.0 (can't just say I love new technology...)? Can thing just work as it is without any changing?... etc (sadly, that's the enterprise world I'm dealing with)

**Comment by wfurt** [comment]
Sorry for the troubles @mrducnguyen, I understand. Not that it would change much for you but I'll check if we can document this is breaking change. 
I was thinking about this for a while (and few other cases) when it may be useful to probe SPN and fall-back if the matching SPN is not found. In that case we can try to right SPN with port and possibly fall-back to SPN without it like we used to do (and as curl does) But that code does not exist and it is unlikely IMHO we would get permissions for servicing 5.0.

**Comment by iSazonov** [comment]
The KB908209 article was removed but there are references in Internet like https://cloudblogs.microsoft.com/dynamics365/no-audience/2011/02/11/kerberos-authentication-does-not-work-when-enterprise-portal-uses-a-non-standard-port/
It seems it says current .Net behavior is correct and a fallback would be wrong. It seems that the target service itself should determine how it accepts authentication.

**Comment by karelz** [comment]
Triage: This seems to be according to the spec. The fact 3.1 was working was a bug. We will document it as breaking change.

**Comment by mattpwhite** [comment]
> It seems that the target service itself should determine how it accepts authentication.

This misunderstands the issue. The sequence is a 401 from the server to client saying it supports Negotiate. There is no facility for the server to tell the client which SPN it should acquire a ticket for, that is left to the client. How that happens varies a bit across platforms (in SSPI applications work directly an SPN, in GSSAPI they usually work a more abstract form) and to some extent applications (the client app has to know the type of service it's trying to access). The result of that in an SPN sent in a TGS-REQ to a KDC, assuming there is no ticket in the cache (and no negative entry). The KDC returns a service ticket, a referral or an error. The client then repeats its HTTP request providing that ticket and providing its knowledge of the session key contained within. In the event that the ticket acquisition fails, Windows will silently fall back to NTLM unless you've disabled it. The point is the only information the server ever gets to tell the client is "I support Negotiate authentication", all other communication is between the client and KDC. 

> Triage: This seems to be according to the spec. The fact 3.1 was working was a bug. We will document it as breaking change.

Which spec says this? There is no mention of port numbers in https://datatracker.ietf.org/doc/html/rfc4559 and what is  implemented in .NET 5.0 is contrary to the behavior of .NET Framework, previous versions of .NET Core, every major browser and curl. At least contrary to the default behavior without custom policies applied. Chrome and IE have support for including the port if a non-standard policy is applied.

**Comment by wfurt** [comment]
any thoughts on this @SteveSyfuhs?

**Comment by SteveSyfuhs** [comment]
There's no good answer here. In an ideal world the client already knows the SPN and doesn't glean it from the URL.

> The point is the only information the server ever gets to tell the client is "I support Negotiate authentication", all other communication is between the client and KDC.

This is actually the critical bit of information here. This is done for security reasons, otherwise an attacker-controlled service can just specify an SPN other than the service you connected to, and do all manner of evil things (this protection _must_ assume no TLS).

I'm not aware of any IETF spec that explicitly states a port is required in the SPN or not. It's happily underspecified, leading to this mess. The port _should_ be on the SPN if it's a non-standard service port. That's important for all sorts of high-density scenarios. However, because SPNs aren't easily modified in AD (i.e. requiring admins), the application should have the option to toggle this behavior off. As I recall we collectively landed on not including the port _by default_.

**Comment by mattpwhite** [comment]
I agree that this is under-specified. The whole thing is a retcon of IE behavior that was only later turned into an RFC. In the absence of a spec stating that the port should be included, the precedent set by existing implementations should be the default behavior.

A facility to alter that default makes some amount of sense if you want to offer it. Any environment using ports in SPNs for HTTP services has already opted themselves into turning knobs in their browsers, curl and other things. Having to opt in to get that behavior in .NET seems reasonable.

Re:density - there are a number of ways to solve that. Additional A records are one of the more traditional ones.

**Comment by NathanielACollier** [comment]
How does .net 5 determine the SPN is not right.  We've been working on setting SPN with port.

We have an existing computer account in AD with lots of SPN.  It has
HTTP/host.domain.com

We created an AD User and gave it the SPN of
HTTP/host.domain.com:8251

8251 is the port where the webservice is at I'm trying to call from .net 5.  .net 5 still send NTLM instead of Negotiate like the server requested.

**Comment by wfurt** [comment]
is `host.domain.com' A record @NathanielACollier ? You can enable tracing and you should see what SPN is being used. 

I also look at the code again and if `Host:` header is set we would use it as is and we would bypass the logic that tries to calculate SPN using port.

**Comment by mattpwhite** [comment]
The automatic downgrade to NTLM is a Windows/SSPI behavior rather than something being done explicitly by .NET. The only thing .NET really gets to choose here is the SPN it passes to SSPI. From that point on it’s basically just shuffling blobs between buffers returned from SSPI APIs and HTTP headers. .NET could examine the attributes of an established context, and that would imply whether the SPN it used was “correct” I guess, but it’s not clear what it would do with that information.

**Comment by SteveSyfuhs** [comment]
I would advise against applying any logic that checks on which protocol gets negotiated, except maybe to optionally fail unsafe protocols like NTLM. Any allow-only list is going to hurt in the long run when we add another nego protocol. That said, any such prevention should happen before the negotiation happens. There's existing Windows policies that disable NTLM as mentioned previously.

**Comment by mattpwhite** [comment]
> I would advise against applying any logic that checks on which protocol gets negotiated

Agreed - whether NTLM is "OK" for someone's environment is better controlled at the OS level rather than within an application. The GP options for auditing/disabling NTLM are fairly good/granular. Using negotiate at all implies some amount of cooperation between app developers and security/infra teams that typically manage those.

> I also look at the code again and if Host: header is set we would use it as is and we would bypass the logic that tries to calculate SPN using port.

That is useful to know, but not always a viable workaround to restore the Core 3.1/Framework behavior. In many scenarios there are layers of libraries and abstractions that make it cumbersome/impractical to set custom values for headers. Even where it is possible to do so, it requires more invasive changes than one might hope for when rebuilding a codebase against a newer version of .NET.

My contention is that:
* In the absence of any spec that says otherwise, the 3.1/Framework behavior should be restored. There does not appear to be a clear justification for the breaking change and reverting would bring .NET back in line with curl and browsers regarding this behavior.
* If an opt-in for ports in SPNs is desired, it should be a new property on `HttpClientHandler` or an opt in global switch for the app or something along those lines.

**Comment by NathanielACollier** [comment]
> is `host.domain.com' A record @NathanielACollier ? You can enable tracing and you should see what SPN is being used.
> 
> I also look at the code again and if `Host:` header is set we would use it as is and we would bypass the logic that tries to calculate SPN using port.

host.domain.com is an Alias.  Is there a guide on enabling tracing?  I've had trouble figuring that out.  I came up with a way to write EventSource to a csv file.  I've listended for these, but don't see anything about SPN in there

                "System.Net",
                "System.Net.Http",
                "System.Net.Sockets"

**Comment by mattpwhite** [comment]
> host.domain.com is an Alias. I think It's a CNAME record

Kerberos clients typically canonicalize hostnames before forming the SPN. Canonicalize meaning that CNAMEs are resolved and the name in the A they point to is used, and that any DNS suffix that was appended to resolve the name is included. Because SSPI expects the client to provide the SPN as a literal string, this canonicalization is an application behavior/convention on Windows. In GSSAPI, where forming the actual SPN is done by the GSSAPI implementation, this is generally controlled by system-level configuration (e.g. in MIT `dns_canonicalize_hostname` in `krb5.conf`, which defaults to true (see [here](https://web.mit.edu/kerberos/krb5-1.12/doc/admin/conf_files/krb5_conf.html)).

Since in SSPI, applications are more free to deviate, they sometimes do. The SMB client in more recent versions of Windows for instance explicitly does *not* canonicalize hostnames before forming an SPN. All major Kerberized HTTP implementations that I am aware of do canonicalize hostnames in SPNs. Chrome accidentally stopped doing this a few years back and quickly fixed the bug to restore the canonicalizing behavior, see the [bug here](https://bugs.chromium.org/p/chromium/issues/detail?id=872665). Once in a while, you'll see something that was primarily developed for Unix systems, where canonicalization was done automatically by their GSSAPI library, ported to Windows, where it's now the responsibility of the application/HTTP library. In those cases, not canonicalizing seems to be an oversight rather than intentional behavior.

Anyway, regarding canonicalization, .NET appears to be in line with other implementations. The inclusion of the port is something else, but the host name portion is correct. The SPN of the server principal should be `HTTP/fqdn`, where `fqdn` is the canonicalized FQDN of the server, not a CNAME.

> Is there a guide on enabling tracing

If you want to know "what" it's doing, a packet capture (ports 53 and 88, UDP and TCP, plus your HTTP traffic if you want that) is generally the easiest way to see, at least on Windows. Just be sure to `klist purge` and `ipconfig /flushdns` first. If you fail to do this, various positive and negative caches will be in play and you may not actually see the relevant Kerberos and DNS traffic on the wire. If you want to know "why", then you'd need tracing on the .NET side, which I'll leave to others. On Unix (at least with MIT), setting the env var `KRB5_TRACE=/dev/stderr` and then running your reproducer is a quick and easy way to trace Kerberos activity.

**Comment by GraffC** [comment]
I can confirm this new behavior that a port is added to the service principal requested by Kerberos in .NET 5 (5.0.7). The same codebase which performs a simple web request with the usage of System.Net.Http "generates" on the same infrastructure an different SPN than .net core 3.1 does.
I can reproduce this behavior and can provide some trace or whatever logs if required (i have the same scenario as described in #51980).
The difference between .NET Core 3.1 and .NET 5 regarding this particular topic for Kerberos Auth is that you can find in the Kerberos-Trace Log files that a port is added for the requested SPNs in .NET 5. 
The Kerberos-Trace Log files differs mainly in the following lines:

- .NET Core 3.1
[10] 1623910720.152732: Requesting tickets for HTTP/internal-somethingon.amazonaws.com@DOMAIN.GLOBAL, referrals on

- .NET 5
[8] 1623910812.651196: Requesting tickets for HTTP/internal-somethingon.amazonaws.com:**7047**@DOMAIN.GLOBAL, referrals on



I got the following questions and would be helpful if you can support here: 

1) Will the .Net Core 3.1 behavior will be restored like @mattpwhite mentioned. 
And if yes, in which version of .NET Framework will this be changed?
	Is there any preview somehow available which can be used?

2) Is there a known workaround you can share, besides adding all the specific port SPNs in the AD (which would confirmed work but lead us into of list of 100+ and more records)?

**Comment by wfurt** [comment]
Do you have any tests from .NET Framework @GraffC?

**Comment by GraffC** [comment]
I can reproduce this "issue" in my environment either for .NET Core 3.1. or .NET 5 but not .NET Full Framework because my App runs on an Linux environment. Our Team would like to migrate our App from .NET Core 3.1 to .NET 5 in order to be more compatible with .NET 6 later on. To be honest we don't want to "stress" our IT department while adding now 100+ new SPNs to our company AD because of the framework update :-).

What can I provide to support the solution finding here and get answers to my questions mentioned above?

**Comment by mattpwhite** [comment]
.NET Framework does not include the port (nor does 3.1). We came across this porting code to .NET Core and it has stalled those efforts indefinitely. Adding different SPNs just for .NET Core when nothing else uses the port is just a non-starter.

**Comment by mattpwhite** [comment]
Any update on this issue? I understand the window for changes in .NET 6 is likely to close soon, and it would be extremely unfortunate for an LTS release to ship with this issue/no clear workaround. Are you willing to accept a PR to revert to the .NET Core 3.1 and .NET Framework behavior?

**Comment by karelz** [comment]
@mattpwhite we still plan to address it in 6.0 - we are considering introducing a switch to opt-in/opt-out -- details are still TBD.
@wfurt do you know what the design would be and if there is something @mattpwhite might helps us with?

**Comment by wfurt** [comment]
Since this is somewhat ambiguous my plan is to add context switch so application can choose the behavior. That it self should not be too difficult. The major obstacle at this point is building test environment and verification.

**Comment by mconnew** [comment]
I wanted to add an additional finding that I discovered. This broke a WCF test scenario so I used the work around of specifying the Host header to override what SPN is used. I tried to capture the request through fiddler, but proxy servers use the Host header to forward the request. So if using a proxy server, the mechanism provided to override the SPN won't work as it causes the proxy server to connect on the default port. So basically .NET 5 can't be used with a server on a non-default port with Kerberos authentication through a proxy server. I know proxies to HTTPS servers use SSL tunneling, but Kerberos is also valid for HTTP. With the .NET 3.1 behavior, if the port number was needed in the SPN, adding it via the Host header wouldn't break anything. There's also the possibility a server implementation might pay too much attention to the Host header and misdirect the request, especially with a reverse proxy frontend.

Is there any news on doing something for issue #25320? WCF recently implemented WSFederation and we're going to be adding a WSTrustChannelFactory very soon. What this means is we're going to be used more for making requests to ADFS servers requiring Kerberos authentication, and [by default](https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2012-R2-and-2012/dn554247(v=ws.11)) HOST and not HTTP is used for the service name part of the SPN that an ADFS server is using. I know HTTP is usually an alias for HOST, but it doesn't have to be. The Host header based fudge that exists is insufficient for these scenarios. And we still can't use UPN's.

**Comment by wfurt** [comment]
I did more testing with IIS & 4.8. (targeting 4.7)
I set up IIS on non-default port and tried to connect with .NET Framework. 
And it works without sending port in SPN. 

Since there is still some ambiguity I added code to decide with AppContext & ENV and I set it off by default e.g. no port to match my observed behavior. (we probably should have done more investigation for #36644) 
Who ever needs it can set either `System.Net.Http.UsePortInSpn` or `DOTNET_SYSTEM_NET_HTTP_USEPORTINSPN` and get the 5.0 behavior. 

To me the big question is what should be do for 5.0. 
Since 6.0 is near by and becoming LTS we can choose to do nothing. 
I'm not sure if we want to change the behavior as bug fix and break anybody who updated SPNs already. 
One possible option would be to port the switch and leave it off to match 5.0 while giving option to choose. 

I hope this would work for you as well @mconnew. 
I would _love_ to progress #25320 and get out of the guessing business. 
It seems like the next step is you finalize API proposal. 
I don't think the actual work would be too difficult.

**Comment by mconnew** [comment]
@wfurt, there are many ways to solve this, and I'm not familiar with the networking teams guidelines on designing new api's so I don't know which of the many options would even be acceptable, never mind preferred. The problem I have is WCF is still targeting .NET Standard 2.0 and I don't believe it's the right thing to do to pull up everyone's requirements to .NET 6/7 to support this. So looking at the current api surface on HttpClientHandler, I can see 3 potential ways to solve this, and none of them may be acceptable. I'll update the bug with my suggestions.

**Comment by karelz** [comment]
Will anyone be able to validate the fix in their environment on 6.0 builds?
Once we have validation, we can initiate backport to 5.0 servicing -- tracked by #57818 -- to unblock affected customers.

**Comment by mconnew** [comment]
@karelz, I already have a vendor working on validating the fix for WCF using a recent 6.0 build. I'll update this issue to confirm once it's been validated.

**Comment by karelz** [comment]
Great, thanks @mconnew!

Another, more important question: Is there someone who cannot upgrade to 6.0 to get this fix? If not, why?
